dot/parachain/availability-store/availability_store.go

```diff
@@ -456,32 +479,250 @@ func (av *AvailabilityStoreSubsystem) processMessages() {
 				}
 
 			case parachaintypes.ActiveLeavesUpdateSignal:
-				av.ProcessActiveLeavesUpdateSignal()
+				err := av.ProcessActiveLeavesUpdateSignal(msg)
+				if err != nil {
+					logger.Errorf("failed to process active leaves update signal: %w", err)
+				}
 
 			case parachaintypes.BlockFinalizedSignal:
-				av.ProcessBlockFinalizedSignal()
+				err := av.ProcessBlockFinalizedSignal(msg)
+				if err != nil {
+					logger.Errorf("failed to process block finalized signal: %w", err)
+				}
 
 			default:
-				logger.Error(parachaintypes.ErrUnknownOverseerMessage.Error())
+				if msg != nil {
+					logger.Infof("unknown message type %T", msg)
+					logger.Error(parachaintypes.ErrUnknownOverseerMessage.Error())
+					// this error shouldn't happen, so we'll panic to catch it during development
+					panic(parachaintypes.ErrUnknownOverseerMessage.Error())
+				}
 			}
-
 		case <-av.ctx.Done():
 			if err := av.ctx.Err(); err != nil {
-				logger.Errorf("ctx error: %v\n", err)
+				logger.Errorf("ctx error: %v", err)
 			}
 			av.wg.Done()
 			return
+		case <-time.After(av.pruningConfig.PruningInterval):
+			av.pruneAll()
+		}
+	}
+}
+
+func (av *AvailabilityStoreSubsystem) ProcessActiveLeavesUpdateSignal(signal parachaintypes.
+	ActiveLeavesUpdateSignal) error {
+	now := timeNow()
+	logger.Infof("ProcessActiveLeavesUpdateSignal %s", signal)
+
+	respChan := make(chan any)
+	message := util.ChainAPIMessage[util.BlockHeader]{
+		Message: util.BlockHeader{
+			Hash: signal.Activated.Hash,
+		},
+		ResponseChannel: respChan,
+	}
+	response, err := util.Call(av.SubSystemToOverseer, message, message.ResponseChannel)
+	if err != nil {
+		return fmt.Errorf("sending message to get block header: %w", err)
+	}
+
+	newBlocks, err := util.DetermineNewBlocks(av.SubSystemToOverseer, av.knownBlocks.isKnown, signal.Activated.Hash,
+		response.(types.Header),
+		av.finalizedNumber)
+	if err != nil {
+		return fmt.Errorf("failed to determine new blocks: %w", err)
+	}
+
+	for _, v := range newBlocks {
+		// start db batch
+		tx := newAvailabilityStoreBatch(&av.availabilityStore)
+
+		err := av.processNewHead(tx, v.Hash, now, v.Header)
+		if err != nil {
+			return fmt.Errorf("failed to process new head: %w", err)
+		}
+
+		// add to known blocks
+		av.knownBlocks.insert(v.Hash, parachaintypes.BlockNumber(v.Header.Number))
+
+		// end db batch
+		err = tx.flush()
+		if err != nil {
+			return fmt.Errorf("failed to flush tx: %w", err)
+		}
+	}
+	return nil
+}
+
+func (av *AvailabilityStoreSubsystem) processNewHead(tx *availabilityStoreBatch, hash common.Hash, now BETimestamp,
+	header types.Header) error {
+	logger.Infof("processNewHead hash %s, now %v, header %v\n", hash, now, header)
+	// TODO: call requestValidators to determine number of validators
```

Comment:
Is this done now?  If not should we create an issue?

Reply:
created issue.